presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -470,6 +481,43 @@ public CounterStat getHighMemorySplitSource()
         return concat(partitionBatches);
     }
 
+    private static Optional<Set<HiveColumnHandle>> mergeRequestedAndPredicateColumns(Optional<Set<HiveColumnHandle>> requestedColumns, Set<HiveColumnHandle> predicateColumns)
+    {
+        if (!requestedColumns.isPresent() || predicateColumns.isEmpty()) {
+            return requestedColumns;
+        }
+
+        Map<String, Set<HiveColumnHandle>> nameToColumnHandles = new HashMap<>();
+        Stream.concat(requestedColumns.get().stream(), predicateColumns.stream())
+                .filter(column -> column.getColumnType() == REGULAR)
+                .forEach(column -> nameToColumnHandles.computeIfAbsent(column.getName(), (key) -> new HashSet<>()).add(column));
+
+        return Optional.of(nameToColumnHandles.values().stream()
+                .map(columnHandles -> {
+                    HiveColumnHandle finalHandle = null;
+                    for (HiveColumnHandle columnHandle : columnHandles) {
+                        if (finalHandle == null) {
```

Comment:
Can this whole section be simplified to
```
if (finalHandle == null || finalHandle.getRequiredSubfields.isEmpty() || columnHandle.getRequiredSubfields.isEmpty()) {
   finalHandle = columnHandle;
}
else {
   finalHandle = (HiveColumnHandle) finalHandle.withRequiredSubfields(ImmutableList.copyOf(ImmutableSet.copyOf(
                                    ImmutableList.<Subfield>builder().addAll(finalHandle.getRequiredSubfields()).addAll(columnHandle.getRequiredSubfields()).build())));
}
```

Also, can we do it as a reduce on the stream instead of having a forloop inside the stream? (if we use groupingBy, don't even need to collect to map first)

Reply:
Thanks for advise on grouping by and reduce - this has made the code a lot simpler now.